dartagnan/src/test/java/com/dat3m/dartagnan/compilation/AbstractCompilationTest.java

```diff
@@ -0,0 +1,147 @@
+        int n = ResourceHelper.LITMUS_RESOURCE_PATH.length();
+        Map<String, Result> expectationMap = ResourceHelper.getExpectedResults();
+
+        try (Stream<Path> fileStream = Files.walk(Paths.get(ResourceHelper.LITMUS_RESOURCE_PATH + litmusPath))) {
+            return fileStream
+                    .filter(Files::isRegularFile)
+                    .map(Path::toString)
+                    .filter(f -> f.endsWith("litmus"))
+                    .filter(f -> expectationMap.containsKey(f.substring(n)))
+                    .collect(ArrayList::new,
+                            (l, f) -> l.add(new Object[]{f}), ArrayList::addAll);
+        }
+    }
+
+
+    // =================== Modifiable behavior ====================
+
+    protected abstract Provider<Arch> getSourceProvider();
+    protected abstract Provider<Arch> getTargetProvider();
+
+    protected Provider<Integer> getTimeoutProvider() {
+        return Provider.fromSupplier(() -> 0);
+    }
+
+    protected long getTimeout() { return 10000; }
+
+    // ============================================================
+
+    protected final Provider<ShutdownManager> shutdownManagerProvider = Provider.fromSupplier(ShutdownManager::create);
+    protected final Provider<Arch> sourceProvider = getSourceProvider();
+    protected final Provider<Arch> targetProvider = getTargetProvider();
+    protected final Provider<String> filePathProvider = () -> path;
+    protected final Provider<Integer> timeoutProvider = getTimeoutProvider();
+    protected final Provider<Program> program1Provider = Providers.createProgramFromPath(filePathProvider);
+    protected final Provider<Program> program2Provider = Providers.createProgramFromPath(filePathProvider);
+    protected final Provider<Wmm> wmm1Provider = Providers.createWmmFromArch(getSourceProvider());
+    protected final Provider<Wmm> wmm2Provider = Providers.createWmmFromArch(getTargetProvider());
+    protected final Provider<EnumSet<Property>> propertyProvider = Provider.fromSupplier(() -> EnumSet.of(Property.getDefault()));
+    protected final Provider<VerificationTask> task1Provider = Provider.fromSupplier(() ->
+        VerificationTask.builder()
+        .withConfig(Configuration.builder()
+            .setOption(INITIALIZE_REGISTERS,String.valueOf(DO_INITIALIZE_REGISTERS))
+            .build())
+        .withTarget(sourceProvider.get())
+        .withSolverTimeout(timeoutProvider.get())
+        .build(program1Provider.get(), wmm1Provider.get(), propertyProvider.get()));
+    protected final Provider<VerificationTask> task2Provider = Provider.fromSupplier(() ->
+    	VerificationTask.builder()
+    	.withConfig(Configuration.builder()
+    		.setOption(INITIALIZE_REGISTERS,String.valueOf(DO_INITIALIZE_REGISTERS))
+    		.build())
+    	.withTarget(targetProvider.get())
+    	.withSolverTimeout(timeoutProvider.get())
+    	.build(program2Provider.get(), wmm2Provider.get(), propertyProvider.get()));
```

Comment:
It might be better to have an explicit method that creates the Provider instead of 2 lambdas.
Possibly something like
```
protected Provider<VerificationTask> createTaskProvider(Provider<Program> progProvider, Provider<Wmm> wmmProvider, Provider<Arch> targetProvider) {  ...  }
```
You could then call this method twice, using the appropriate other providers.